src/hash.h

```diff
@@ -160,6 +160,42 @@ class CHashWriter
     }
 };
 
+/** Reads data from an underlying stream, while hashing the read data. */
+template<typename Source>
+class CHashVerifier : public CHashWriter
+{
+private:
+    Source* source;
+
+public:
+    CHashVerifier(Source* source_) : CHashWriter(source_->GetType(), source_->GetVersion()), source(source_) {}
+
+    void read(char* pch, size_t nSize)
+    {
```

Comment:
It seems that the invoked read and write method already need to deal with nSize==0 anyway?